<h1 style='background-color: Indigo; padding: 10px; color: white; text-align: center;'>DATA CLEANING </h1>

First things first, we will import all the required packages

In [243]:
import pandas as pd
import numpy as np
from datetime import datetime, timedelta

We will start by cleaning all the sales data for each of the three websites

<h1 style='background-color: Indigo; padding: 10px; color: white'> Sales data </h1>

We will import the sales data for the three websites

In [244]:
# Rightmove
rm = pd.read_csv('RightMove_Sales_data.csv')

# On the Market
om = pd.read_csv('OnTheMarket_Sales_data.csv')

#Zoopla
zp = pd.read_csv('Zoopla_Sales_data.csv')

We will remove the unwanted column 'Unnamed from all the datasets'

In [245]:
# Rightmove
rm = rm.drop('Unnamed: 0', axis=1)

# On the Market
om = om.drop('Unnamed: 0', axis=1)

#Zoopla
zp = zp.drop('Unnamed: 0', axis=1)

# Rightmove

Now we will assess and clean the sales data for rightmove

## a) Data Assesment

In [246]:
rm.head(10)

Unique_Id Location Tansaction_Type Property_Type   
0  BR1S0000RM      BR1           Sales      Terraced  \
1  BR1S0001RM      BR1           Sales      Detached   
2  BR1S0002RM      BR1           Sales      Detached   
3  BR1S0003RM      BR1           Sales      Detached   
4  BR1S0004RM      BR1           Sales         House   
5  BR1S0005RM      BR1           Sales      Detached   
6  BR1S0006RM      BR1           Sales      Detached   
7  BR1S0007RM      BR1           Sales      Detached   
8  BR1S0008RM      BR1           Sales      Detached   
9  BR1S0009RM      BR1           Sales         House   

                                             Address  Bedrooms  Bathrooms   
0                          Palace View, Bromley, BR1         3          1  \
1  Provident House, 122-126 High Street, Bromley BR1         6          7   
2                           Woodlands Close, Bromley         5          5   
3                       Sundridge Avenue Bromley BR1         6          3   
4               Mavelstone Close, Bromley, Kent, BR1         8          3   
5                         Southborough Road, Bromley         5          4   
6              Bickley Park Road, Bickley, Kent, BR1         5          5   
7                          Timms Close, Bickley, BR1         5          1   
8                    Timms Close, Bromley, Kent, BR1         4          5   
9                             Westbury Road, Bromley         5          3   

        Price                                        Description   
0    £745,000  A fantastic opportunity to purchase this mid t...  \
1  £5,750,000  Extremely rare opportunity to purchase this tr...   
2  £3,500,000  Guide Price £2,500,000 to £2,600,000. Brand ne...   
3  £2,500,000  A fantastic family home offered in quiet sough...   
4  £2,400,000  Of interest to property developers, speculator...   
5  £2,300,000  An attractively designed and built modern hous...   
6  £2,250,000  Unique, custom designed 'Green' award winning ...   
7  £2,250,000  Guide Price £2,250,000 - £2,500,000 A unique a...   
8  £2,250,000  Set on a generous plot with a stunning 127 ft ...   
9  £1,850,000  Woodlands Road is probably one of Chislehurst'...   

            Listing_Date                                              Agent   
0    Added on 27/03/2023                                  Allow all cookies  \
1                    NaN  Added on 27/03/2023 by Truepenny's Property Co...   
2    Added on 25/03/2023              Marketed by Knight Frank Land, London   
3    Added on 16/03/2023   Added on 25/03/2023 by Alan de Maid, Chislehurst   
4    Added on 22/02/2023  Added on 16/03/2023 by Unique, Langford Russel...   
5  Reduced on 13/04/2023            Added on 22/02/2023 by jdm, Chislehurst   
6    Added on 16/04/2023          Reduced on 13/04/2023 by Hunters, Bromley   
7  Reduced on 27/04/2023            Added on 16/04/2023 by jdm, Chislehurst   
8  Reduced on 27/04/2023   Reduced on 27/04/2023 by Fine & Country, Bromley   
9    Added on 15/03/2023  Reduced on 27/04/2023 by Sinclair Hammelton, B...   

  Listing_Source                                        listing_URL  
0      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
1      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
2      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
3      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
4      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
5      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
6      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
7      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
8      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
9      RightMove  https://www.rightmove.co.uk/property-for-sale/...

In [247]:
# Checking the shape of the data
rm.shape

(49641, 13)

In [248]:
# Checking the data distribution
rm.describe()

Bedrooms     Bathrooms
count  49641.000000  49641.000000
mean       2.505933      1.648617
std        1.316835      1.138742
min        1.000000      1.000000
25%        2.000000      1.000000
50%        2.000000      1.000000
75%        3.000000      2.000000
max       28.000000    144.000000

In [249]:
# Assessing the data types
rm.dtypes

Unique_Id          object
Location           object
Tansaction_Type    object
Property_Type      object
Address            object
Bedrooms            int64
Bathrooms           int64
Price              object
Description        object
Listing_Date       object
Agent              object
Listing_Source     object
listing_URL        object
dtype: object

## b) Dealing with nulls

In [250]:
rm.isnull().sum()

Unique_Id            0
Location             0
Tansaction_Type      0
Property_Type        1
Address              0
Bedrooms             0
Bathrooms            0
Price              293
Description          4
Listing_Date       100
Agent                0
Listing_Source       0
listing_URL          0
dtype: int64

Three of the column has nulls

### i) Dealing with nulls in the Property_Type column

In [251]:
# For the Property_type column
rm[rm['Property_Type'].isnull()]

Unique_Id Location Tansaction_Type Property_Type   
21919  N3S00021919RM       N3           Sales           NaN  \

                                Address  Bedrooms  Bathrooms     Price   
21919  Edge Hill Avenue, London, N3 3AY         2          2  £405,000  \

                                             Description         Listing_Date   
21919  New beginnings in North LondonLocated in Finch...  Added on 03/03/2023  \

                                                 Agent Listing_Source   
21919  Added on 03/02/2023 by Martyn Gerrard, Finchley      RightMove  \

                                             listing_URL  
21919  https://www.rightmove.co.uk/property-for-sale/...

We will replace the null with 'Not listed'

In [252]:
rm['Property_Type'].fillna('Not listed', inplace=True)

### ii) Dealing with nulls in the Price column

In [253]:
# For the Price column
rm[rm['Price'].isnull()].head()

Unique_Id Location Tansaction_Type   Property_Type   
648  BR3S000648RM      BR3           Sales        Detached  \
682  BR3S000682RM      BR3           Sales        Detached   
684  BR3S000684RM      BR3           Sales        Terraced   
686  BR3S000686RM      BR3           Sales   Semi-Detached   
691  BR3S000691RM      BR3           Sales  End of Terrace   

                                             Address  Bedrooms  Bathrooms   
648                       Edgefield Close, Beckenham         5          2  \
682  Langley Court, 11 Roman Way, Beckenham, BR3 3FH         3          1   
684                   St. Martins Lane Beckenham BR3         4          2   
686                        Roman Way, Beckenham, BR3         4          2   
691                   Albemarle Road, Beckenham, BR3         3          2   

    Price                                        Description   
648   NaN  Boasting a peaceful gated cul-de-sac position,...  \
682   NaN  Introducing the Miller, a 3 bedroom link-detac...   
684   NaN  **Guide Price £900,000 to £950,000** This four...   
686   NaN  Neighbouring leafy Park Langley, Cala Home's L...   
691   NaN  This charming 3 bedroom end of terrace house i...   

                  Listing_Date                                  Agent   
648  Langley Court Development  NEW HOME\r\nLangley Court Development  \
682  Langley Court Development  NEW HOME\r\nLangley Court Development   
684                   NEW HOME  NEW HOME\r\nLangley Court Development   
686        Added on 16/11/2022  NEW HOME\r\nLangley Court Development   
691      Reduced on 15/02/2023  Reduced on 18/04/2023 by jdm, Bromley   

    Listing_Source                                        listing_URL  
648      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
682      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
684      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
686      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
691      RightMove  https://www.rightmove.co.uk/property-for-sale/...

We will replace the null with 'Not listed'

In [254]:
rm['Price'].fillna('Not listed', inplace=True)

### iii) Dealing with nulls in the Listing_Date column

In [255]:
# For the Listing Date column
rm[rm['Listing_Date'].isnull()].head()

Unique_Id Location Tansaction_Type  Property_Type   
1        BR1S0001RM      BR1           Sales       Detached  \
148    BR1S000148RM      BR1           Sales  Land for sale   
603    BR2S000603RM      BR2           Sales           Flat   
1034  BR5S0001034RM      BR5           Sales  Plot for sale   
1277  BR6S0001277RM      BR6           Sales  Plot for sale   

                                                Address  Bedrooms  Bathrooms   
1     Provident House, 122-126 High Street, Bromley BR1         6          7  \
148                    College Road, Bromley, Kent, BR1         2          2   
603                        Homesdale Road, Bromley, BR2         1          1   
1034                              Rosecroft Social Club         5          3   
1277               Land At Charterhouse Road, Orpington         4          4   

           Price                                        Description   
1     £5,750,000  Extremely rare opportunity to purchase this tr...  \
148     £525,000  Freehold Victorian House Planning granted for ...   
603     £275,000  **GUIDE PRICE £275,000 - £300,000** MANN are d...   
1034  £1,250,000  PLANNING APPROVED FOR DEVELOPMENT OF 4X FLATS ...   
1277  £1,800,000                            Development Opportunity   

     Listing_Date                                              Agent   
1             NaN  Added on 27/03/2023 by Truepenny's Property Co...  \
148           NaN  Added on 07/04/2023 by Kinleigh Folkard & Hayw...   
603           NaN     Added on 17/04/2023 by Fish Need Water, London   
1034          NaN        Added on 25/04/2023 by Proctors, Petts Wood   
1277          NaN  Reduced on 27/02/2023 by Alan de Maid, Locksbo...   

     Listing_Source                                        listing_URL  
1         RightMove  https://www.rightmove.co.uk/property-for-sale/...  
148       RightMove  https://www.rightmove.co.uk/property-for-sale/...  
603       RightMove  https://www.rightmove.co.uk/property-for-sale/...  
1034      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
1277      RightMove  https://www.rightmove.co.uk/property-for-sale/...

We will replace the null with 'Not listed'

In [256]:
rm['Listing_Date'].fillna('Not listed', inplace=True)

### iv) Dealing with nulls in the Listing_Date column

In [257]:
# For the Description column
rm[rm['Description'].isnull()].head()

Unique_Id Location Tansaction_Type Property_Type   
45663  W2S00045663RM       W2           Sales          Flat  \
46045  W2S00046045RM       W2           Sales     Apartment   
46056  W2S00046056RM       W2           Sales          Flat   
46127  W2S00046127RM       W2           Sales          Flat   

                                                 Address  Bedrooms  Bathrooms   
45663   Flat 9, 69-71 Gloucester Terrace, London, W2 3DH         1          3  \
46045                         Monmouth Place, London, W2         1          1   
46056  Flat 15, 69-71 Gloucester Terrace, London, W2 3DH         1          2   
46127   Flat 9, 69-71 Gloucester Terrace, London, W2 3DH         1          1   

          Price Description           Listing_Date   
45663  £240,000         NaN    Added on 05/05/2023  \
46045  £940,000         NaN  Reduced on 23/03/2023   
46056  £240,000         NaN    Added on 05/05/2023   
46127  £240,000         NaN    Added on 05/05/2023   

                                                   Agent Listing_Source   
45663                                  Allow all cookies      RightMove  \
46045  Reduced on 02/05/2023 by Dexters, Hyde Park & ...      RightMove   
46056                                  Allow all cookies      RightMove   
46127                                  Allow all cookies      RightMove   

                                             listing_URL  
45663  https://www.rightmove.co.uk/property-for-sale/...  
46045  https://www.rightmove.co.uk/property-for-sale/...  
46056  https://www.rightmove.co.uk/property-for-sale/...  
46127  https://www.rightmove.co.uk/property-for-sale/...

We will replace the null with 'Not listed'

In [258]:
rm['Description'].fillna('Not listed', inplace=True)

In [259]:
rm.isnull().sum()

Unique_Id          0
Location           0
Tansaction_Type    0
Property_Type      0
Address            0
Bedrooms           0
Bathrooms          0
Price              0
Description        0
Listing_Date       0
Agent              0
Listing_Source     0
listing_URL        0
dtype: int64

All the nulls have been succesfully dealt with.

## c) Data Cleaning

In [260]:
rm.head()

Unique_Id Location Tansaction_Type Property_Type   
0  BR1S0000RM      BR1           Sales      Terraced  \
1  BR1S0001RM      BR1           Sales      Detached   
2  BR1S0002RM      BR1           Sales      Detached   
3  BR1S0003RM      BR1           Sales      Detached   
4  BR1S0004RM      BR1           Sales         House   

                                             Address  Bedrooms  Bathrooms   
0                          Palace View, Bromley, BR1         3          1  \
1  Provident House, 122-126 High Street, Bromley BR1         6          7   
2                           Woodlands Close, Bromley         5          5   
3                       Sundridge Avenue Bromley BR1         6          3   
4               Mavelstone Close, Bromley, Kent, BR1         8          3   

        Price                                        Description   
0    £745,000  A fantastic opportunity to purchase this mid t...  \
1  £5,750,000  Extremely rare opportunity to purchase this tr...   
2  £3,500,000  Guide Price £2,500,000 to £2,600,000. Brand ne...   
3  £2,500,000  A fantastic family home offered in quiet sough...   
4  £2,400,000  Of interest to property developers, speculator...   

          Listing_Date                                              Agent   
0  Added on 27/03/2023                                  Allow all cookies  \
1           Not listed  Added on 27/03/2023 by Truepenny's Property Co...   
2  Added on 25/03/2023              Marketed by Knight Frank Land, London   
3  Added on 16/03/2023   Added on 25/03/2023 by Alan de Maid, Chislehurst   
4  Added on 22/02/2023  Added on 16/03/2023 by Unique, Langford Russel...   

  Listing_Source                                        listing_URL  
0      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
1      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
2      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
3      RightMove  https://www.rightmove.co.uk/property-for-sale/...  
4      RightMove  https://www.rightmove.co.uk/property-for-sale/...

### i) Cleaning the Listing_Date column

Now, we will modify the Listed date column

In [261]:
# Removing the 'Added on' from the values
rm['Listing_Date'] = rm['Listing_Date'].str.replace('Added on ', '', regex=True)
rm['Listing_Date'] = rm['Listing_Date'].str.replace('Reduced on ', '', regex=True)

In [262]:
rm.head()

Unique_Id Location Tansaction_Type Property_Type   
0  BR1S0000RM      BR1           Sales      Terraced  \
1  BR1S0001RM      BR1           Sales      Detached   
2  BR1S0002RM      BR1           Sales      Detached   
3  BR1S0003RM      BR1           Sales      Detached   
4  BR1S0004RM      BR1           Sales         House   

                                             Address  Bedrooms  Bathrooms   
0                          Palace View, Bromley, BR1         3          1  \
1  Provident House, 122-126 High Street, Bromley BR1         6          7   
2                           Woodlands Close, Bromley         5          5   
3                       Sundridge Avenue Bromley BR1         6          3   
4               Mavelstone Close, Bromley, Kent, BR1         8          3   

        Price                                        Description Listing_Date   
0    £745,000  A fantastic opportunity to purchase this mid t...   27/03/2023  \
1  £5,750,000  Extremely rare opportunity to purchase this tr...   Not listed   
2  £3,500,000  Guide Price £2,500,000 to £2,600,000. Brand ne...   25/03/2023   
3  £2,500,000  A fantastic family home offered in quiet sough...   16/03/2023   
4  £2,400,000  Of interest to property developers, speculator...   22/02/2023   

                                               Agent Listing_Source   
0                                  Allow all cookies      RightMove  \
1  Added on 27/03/2023 by Truepenny's Property Co...      RightMove   
2              Marketed by Knight Frank Land, London      RightMove   
3   Added on 25/03/2023 by Alan de Maid, Chislehurst      RightMove   
4  Added on 16/03/2023 by Unique, Langford Russel...      RightMove   

                                         listing_URL  
0  https://www.rightmove.co.uk/property-for-sale/...  
1  https://www.rightmove.co.uk/property-for-sale/...  
2  https://www.rightmove.co.uk/property-for-sale/...  
3  https://www.rightmove.co.uk/property-for-sale/...  
4  https://www.rightmove.co.uk/property-for-sale/...

Convert any non-date values in the Listed_Date column to 'Not Listed'

In [263]:
# Convert the column to string type
rm['Listing_Date'] = rm['Listing_Date'].astype(str)

# Check if the value matches the date format
mask = rm['Listing_Date'].str.match(r'\d{2}/\d{2}/\d{4}')

# Replace non-matching values with 'Not Listed'
rm.loc[~mask, 'Listing_Date'] = 'Not Listed'

In [264]:
rm.Listing_Date.value_counts()

Listing_Date
Not Listed    6252
03/05/2023    1211
02/05/2023    1154
28/04/2023     981
24/04/2023     929
              ... 
26/09/2019       1
01/11/2020       1
30/09/2020       1
26/12/2018       1
16/06/2021       1
Name: count, Length: 929, dtype: int64

### ii) Cleaning the Agent column

Modifying the Agent column

In [265]:
# Removing all strings before the word 'by'
try:
    rm['Agent'] = rm['Agent'].str.split('by', n=1).str[1].str.strip()
except ValueError:
    pass

In [266]:
rm.head()

Unique_Id Location Tansaction_Type Property_Type   
0  BR1S0000RM      BR1           Sales      Terraced  \
1  BR1S0001RM      BR1           Sales      Detached   
2  BR1S0002RM      BR1           Sales      Detached   
3  BR1S0003RM      BR1           Sales      Detached   
4  BR1S0004RM      BR1           Sales         House   

                                             Address  Bedrooms  Bathrooms   
0                          Palace View, Bromley, BR1         3          1  \
1  Provident House, 122-126 High Street, Bromley BR1         6          7   
2                           Woodlands Close, Bromley         5          5   
3                       Sundridge Avenue Bromley BR1         6          3   
4               Mavelstone Close, Bromley, Kent, BR1         8          3   

        Price                                        Description Listing_Date   
0    £745,000  A fantastic opportunity to purchase this mid t...   27/03/2023  \
1  £5,750,000  Extremely rare opportunity to purchase this tr...   Not Listed   
2  £3,500,000  Guide Price £2,500,000 to £2,600,000. Brand ne...   25/03/2023   
3  £2,500,000  A fantastic family home offered in quiet sough...   16/03/2023   
4  £2,400,000  Of interest to property developers, speculator...   22/02/2023   

                                        Agent Listing_Source   
0                                         NaN      RightMove  \
1  Truepenny's Property Consultants, Charlton      RightMove   
2                   Knight Frank Land, London      RightMove   
3                   Alan de Maid, Chislehurst      RightMove   
4            Unique, Langford Russell Bromley      RightMove   

                                         listing_URL  
0  https://www.rightmove.co.uk/property-for-sale/...  
1  https://www.rightmove.co.uk/property-for-sale/...  
2  https://www.rightmove.co.uk/property-for-sale/...  
3  https://www.rightmove.co.uk/property-for-sale/...  
4  https://www.rightmove.co.uk/property-for-sale/...

In [267]:
# Replace all nulls with 'Not listed'
rm['Agent'].fillna('Not listed', inplace=True)

# On the Market

Now we will clean the sales data for On the Market

## a) Data Assessment

In [268]:
om.head()

Unique_Id Location Tansaction_Type   
0  BR1S00001OM      BR1           Sales  \
1  BR1S00001OM      BR1           Sales   
2  BR1S00001OM      BR1           Sales   
3  BR1S00001OM      BR1           Sales   
4  BR1S00001OM      BR1           Sales   

                            Property_Type                             Address   
0       3 bedroom terraced house for sale         Parkside Avenue Bromley BR1  \
1       2 bedroom terraced house for sale    Foxbury Road, Bromley, Kent, BR1   
2  3 bedroom semi-detached house for sale  Bishops Avenue, Bromley, Kent, BR1   
3       2 bedroom terraced house for sale              Gladwell Road, Bromley   
4            1 bedroom apartment for sale     Liddon Road, Bromley, Kent, BR1   

  Bedrooms  Bathrooms     Price   
0        3        1.0  £550,000  \
1        2        1.0  £525,000   
2        3        1.0  £575,000   
3        2        1.0  £475,000   
4        1        1.0  £250,000   

                                         Description           Listing_Date   
0  Parkside Avenue Bromley BR1 3 bedroom terraced...  OnTheMarket > 14 days  \
1  Foxbury Road, Bromley, Kent, BR1 2 bedroom ter...      OnTheMarket today   
2  Bishops Avenue, Bromley, Kent, BR1 3 bedroom s...          Reduced today   
3  Gladwell Road, Bromley. 2 bedroom terraced hou...  OnTheMarket > 14 days   
4  Liddon Road, Bromley, Kent, BR1. 1 bedroom apa...  OnTheMarket < 14 days   

                                        Agent Listing_Source   
0                    Edward Ashdale - Bromley   OnTheMartket  \
1          Sinclair Hammelton - Bromley Sales   OnTheMartket   
2          Sinclair Hammelton - Bromley Sales   OnTheMartket   
3  Kinleigh Folkard & Hayward - Bromley Sales   OnTheMartket   
4          Sinclair Hammelton - Bromley Sales   OnTheMartket   

                                         listing_URL Current_date  
0  https://www.onthemarket.com/for-sale/property/...   15/05/2023  
1  https://www.onthemarket.com/for-sale/property/...   15/05/2023  
2  https://www.onthemarket.com/for-sale/property/...   15/05/2023  
3  https://www.onthemarket.com/for-sale/property/...   15/05/2023  
4  https://www.onthemarket.com/for-sale/property/...   15/05/2023

In [269]:
# Checking the shape of the Dataset
om.shape

(14790, 14)

In [270]:
# Checking the datatype
om.dtypes

Unique_Id           object
Location            object
Tansaction_Type     object
Property_Type       object
Address             object
Bedrooms            object
Bathrooms          float64
Price               object
Description         object
Listing_Date        object
Agent               object
Listing_Source      object
listing_URL         object
Current_date        object
dtype: object

In [271]:
# Checking the shape of the Dataset
om.describe()

Bathrooms
count  13946.000000
mean       1.643769
std        0.965680
min        1.000000
25%        1.000000
50%        1.000000
75%        2.000000
max       15.000000

## b) Dealing with nulls

In [272]:
om.isnull().sum()

Unique_Id              0
Location               0
Tansaction_Type        0
Property_Type          0
Address                0
Bedrooms             107
Bathrooms            844
Price                  0
Description            0
Listing_Date           0
Agent              10216
Listing_Source         0
listing_URL            0
Current_date           0
dtype: int64

Now, we will replace all the nulls with 'Not listed'

In [273]:
# Replacing nulls in the Bedroom column with 'Not Listed'
om['Bedrooms'].fillna('Not listed', inplace=True)

# Replacing nulls in the Bathrooms column with 'Not Listed'
om['Bathrooms'].fillna('Not listed', inplace=True)

# Replacing nulls in the Agent column with 'Not Listed'
om['Agent'].fillna('Not listed', inplace=True)

In [274]:
om.isnull().sum()

Unique_Id          0
Location           0
Tansaction_Type    0
Property_Type      0
Address            0
Bedrooms           0
Bathrooms          0
Price              0
Description        0
Listing_Date       0
Agent              0
Listing_Source     0
listing_URL        0
Current_date       0
dtype: int64

All nulls have been succesfully dealt with

## c) Data Cleaning

In [275]:
om.head()

Unique_Id Location Tansaction_Type   
0  BR1S00001OM      BR1           Sales  \
1  BR1S00001OM      BR1           Sales   
2  BR1S00001OM      BR1           Sales   
3  BR1S00001OM      BR1           Sales   
4  BR1S00001OM      BR1           Sales   

                            Property_Type                             Address   
0       3 bedroom terraced house for sale         Parkside Avenue Bromley BR1  \
1       2 bedroom terraced house for sale    Foxbury Road, Bromley, Kent, BR1   
2  3 bedroom semi-detached house for sale  Bishops Avenue, Bromley, Kent, BR1   
3       2 bedroom terraced house for sale              Gladwell Road, Bromley   
4            1 bedroom apartment for sale     Liddon Road, Bromley, Kent, BR1   

  Bedrooms Bathrooms     Price   
0        3       1.0  £550,000  \
1        2       1.0  £525,000   
2        3       1.0  £575,000   
3        2       1.0  £475,000   
4        1       1.0  £250,000   

                                         Description           Listing_Date   
0  Parkside Avenue Bromley BR1 3 bedroom terraced...  OnTheMarket > 14 days  \
1  Foxbury Road, Bromley, Kent, BR1 2 bedroom ter...      OnTheMarket today   
2  Bishops Avenue, Bromley, Kent, BR1 3 bedroom s...          Reduced today   
3  Gladwell Road, Bromley. 2 bedroom terraced hou...  OnTheMarket > 14 days   
4  Liddon Road, Bromley, Kent, BR1. 1 bedroom apa...  OnTheMarket < 14 days   

                                        Agent Listing_Source   
0                    Edward Ashdale - Bromley   OnTheMartket  \
1          Sinclair Hammelton - Bromley Sales   OnTheMartket   
2          Sinclair Hammelton - Bromley Sales   OnTheMartket   
3  Kinleigh Folkard & Hayward - Bromley Sales   OnTheMartket   
4          Sinclair Hammelton - Bromley Sales   OnTheMartket   

                                         listing_URL Current_date  
0  https://www.onthemarket.com/for-sale/property/...   15/05/2023  
1  https://www.onthemarket.com/for-sale/property/...   15/05/2023  
2  https://www.onthemarket.com/for-sale/property/...   15/05/2023  
3  https://www.onthemarket.com/for-sale/property/...   15/05/2023  
4  https://www.onthemarket.com/for-sale/property/...   15/05/2023

### i) Cleaning the Listing_Date Column

Here are the values available in the Listing_Date

In [276]:
om.Listing_Date.value_counts()

Listing_Date
OnTheMarket > 14 days    11559
OnTheMarket < 14 days      869
OnTheMarket < 7 days       823
OnTheMarket yesterday      423
Reduced < 7 days           423
Reduced < 14 days          385
Reduced yesterday          160
OnTheMarket today          100
Reduced today               44
Reduced > 14 days            4
Name: count, dtype: int64

To ensure uniformity with the dataset from other websites, we need to convert the values into a Date datatype. We will accomplish this by utilizing the 'Current_date' column, which represents the date the data was web scraped, and the 'Listing_date' column to compute a new column that will serve as our standardized listing date.

E.g If current date = 15/05/2023 and Listing date = OnTheMarket > 14 days. Date_main will be  01/05/2023

If current date = 15/05/2023 and Listing date = OnTheMarket today. Date_main will be 15/05/2023

In [277]:
# Convert 'Current_date' column to datetime type
om['Current_date'] = pd.to_datetime(om['Current_date'], format='%d/%m/%Y')

# Calculate the new 'Date_main' column based on different conditions
om['Date_main'] = om.apply(
    lambda row: (
        (row['Current_date']).strftime('%d/%m/%Y')
        if row['Listing_Date'] in ['Reduced today', 'OnTheMarket today']
        else (row['Current_date'] - timedelta(days=1)).strftime('%d/%m/%Y')
        if row['Listing_Date'] in ['Reduced yesterday', 'OnTheMarket yesterday']
        else (row['Current_date'] - timedelta(days=14)).strftime('%d/%m/%Y')  # > (current date - 14 days)
        if row['Listing_Date'] in ['OnTheMarket > 14 days', 'Reduced > 14 days']
        else (row['Current_date'] - timedelta(days=14)).strftime('%d/%m/%Y')  # < (current date - 14 days)
        if row['Listing_Date'] in ['OnTheMarket < 14 days', 'Reduced < 14 days']
        else (row['Current_date'] - timedelta(days=7)).strftime('%d/%m/%Y')  # < (current date - 7 days)
        if row['Listing_Date'] in ['OnTheMarket < 7 days', 'Reduced < 7 days']
        else None  # Add any other conditions or default value if needed
    ),
    axis=1
)

# Convert the column to datetime type
om['Date_main'] = pd.to_datetime(om['Date_main'], format='%d/%m/%Y')

# Convert the values to the desired format
om['Date_main'] = om['Date_main'].dt.strftime('%d/%m/%Y')


In [278]:
om.head()

Unique_Id Location Tansaction_Type   
0  BR1S00001OM      BR1           Sales  \
1  BR1S00001OM      BR1           Sales   
2  BR1S00001OM      BR1           Sales   
3  BR1S00001OM      BR1           Sales   
4  BR1S00001OM      BR1           Sales   

                            Property_Type                             Address   
0       3 bedroom terraced house for sale         Parkside Avenue Bromley BR1  \
1       2 bedroom terraced house for sale    Foxbury Road, Bromley, Kent, BR1   
2  3 bedroom semi-detached house for sale  Bishops Avenue, Bromley, Kent, BR1   
3       2 bedroom terraced house for sale              Gladwell Road, Bromley   
4            1 bedroom apartment for sale     Liddon Road, Bromley, Kent, BR1   

  Bedrooms Bathrooms     Price   
0        3       1.0  £550,000  \
1        2       1.0  £525,000   
2        3       1.0  £575,000   
3        2       1.0  £475,000   
4        1       1.0  £250,000   

                                         Description           Listing_Date   
0  Parkside Avenue Bromley BR1 3 bedroom terraced...  OnTheMarket > 14 days  \
1  Foxbury Road, Bromley, Kent, BR1 2 bedroom ter...      OnTheMarket today   
2  Bishops Avenue, Bromley, Kent, BR1 3 bedroom s...          Reduced today   
3  Gladwell Road, Bromley. 2 bedroom terraced hou...  OnTheMarket > 14 days   
4  Liddon Road, Bromley, Kent, BR1. 1 bedroom apa...  OnTheMarket < 14 days   

                                        Agent Listing_Source   
0                    Edward Ashdale - Bromley   OnTheMartket  \
1          Sinclair Hammelton - Bromley Sales   OnTheMartket   
2          Sinclair Hammelton - Bromley Sales   OnTheMartket   
3  Kinleigh Folkard & Hayward - Bromley Sales   OnTheMartket   
4          Sinclair Hammelton - Bromley Sales   OnTheMartket   

                                         listing_URL Current_date   Date_main  
0  https://www.onthemarket.com/for-sale/property/...   2023-05-15  01/05/2023  
1  https://www.onthemarket.com/for-sale/property/...   2023-05-15  15/05/2023  
2  https://www.onthemarket.com/for-sale/property/...   2023-05-15  15/05/2023  
3  https://www.onthemarket.com/for-sale/property/...   2023-05-15  01/05/2023  
4  https://www.onthemarket.com/for-sale/property/...   2023-05-15  01/05/2023

In [279]:
om.shape

(14790, 15)

In [280]:
# Drop the 'Current_date' and 'Listing_Date' columns
om.drop(['Current_date', 'Listing_Date'], axis=1, inplace=True)

For uniformity with the datsets from other websites, the Date_main column will be renamed as Listing_Date

In [281]:
# Rename the 'Date_main' column to 'Listing_Date'
om.rename(columns={'Date_main': 'Listing_Date'}, inplace=True)

In [282]:
om.head()

Unique_Id Location Tansaction_Type   
0  BR1S00001OM      BR1           Sales  \
1  BR1S00001OM      BR1           Sales   
2  BR1S00001OM      BR1           Sales   
3  BR1S00001OM      BR1           Sales   
4  BR1S00001OM      BR1           Sales   

                            Property_Type                             Address   
0       3 bedroom terraced house for sale         Parkside Avenue Bromley BR1  \
1       2 bedroom terraced house for sale    Foxbury Road, Bromley, Kent, BR1   
2  3 bedroom semi-detached house for sale  Bishops Avenue, Bromley, Kent, BR1   
3       2 bedroom terraced house for sale              Gladwell Road, Bromley   
4            1 bedroom apartment for sale     Liddon Road, Bromley, Kent, BR1   

  Bedrooms Bathrooms     Price   
0        3       1.0  £550,000  \
1        2       1.0  £525,000   
2        3       1.0  £575,000   
3        2       1.0  £475,000   
4        1       1.0  £250,000   

                                         Description   
0  Parkside Avenue Bromley BR1 3 bedroom terraced...  \
1  Foxbury Road, Bromley, Kent, BR1 2 bedroom ter...   
2  Bishops Avenue, Bromley, Kent, BR1 3 bedroom s...   
3  Gladwell Road, Bromley. 2 bedroom terraced hou...   
4  Liddon Road, Bromley, Kent, BR1. 1 bedroom apa...   

                                        Agent Listing_Source   
0                    Edward Ashdale - Bromley   OnTheMartket  \
1          Sinclair Hammelton - Bromley Sales   OnTheMartket   
2          Sinclair Hammelton - Bromley Sales   OnTheMartket   
3  Kinleigh Folkard & Hayward - Bromley Sales   OnTheMartket   
4          Sinclair Hammelton - Bromley Sales   OnTheMartket   

                                         listing_URL Listing_Date  
0  https://www.onthemarket.com/for-sale/property/...   01/05/2023  
1  https://www.onthemarket.com/for-sale/property/...   15/05/2023  
2  https://www.onthemarket.com/for-sale/property/...   15/05/2023  
3  https://www.onthemarket.com/for-sale/property/...   01/05/2023  
4  https://www.onthemarket.com/for-sale/property/...   01/05/2023

Now we will move the new listed date column to it's right position

In [283]:
# Get the column names
columns = om.columns.tolist()

# Specify the position where you want to move the column (e.g., from 4th position to 1st position)
from_position = 12
to_position = 9

# Move the column to the desired position
columns.insert(to_position, columns.pop(from_position))

# Reassign the columns in the DataFrame with the updated order
om = om[columns]



In [284]:
om.head()

Unique_Id Location Tansaction_Type   
0  BR1S00001OM      BR1           Sales  \
1  BR1S00001OM      BR1           Sales   
2  BR1S00001OM      BR1           Sales   
3  BR1S00001OM      BR1           Sales   
4  BR1S00001OM      BR1           Sales   

                            Property_Type                             Address   
0       3 bedroom terraced house for sale         Parkside Avenue Bromley BR1  \
1       2 bedroom terraced house for sale    Foxbury Road, Bromley, Kent, BR1   
2  3 bedroom semi-detached house for sale  Bishops Avenue, Bromley, Kent, BR1   
3       2 bedroom terraced house for sale              Gladwell Road, Bromley   
4            1 bedroom apartment for sale     Liddon Road, Bromley, Kent, BR1   

  Bedrooms Bathrooms     Price   
0        3       1.0  £550,000  \
1        2       1.0  £525,000   
2        3       1.0  £575,000   
3        2       1.0  £475,000   
4        1       1.0  £250,000   

                                         Description Listing_Date   
0  Parkside Avenue Bromley BR1 3 bedroom terraced...   01/05/2023  \
1  Foxbury Road, Bromley, Kent, BR1 2 bedroom ter...   15/05/2023   
2  Bishops Avenue, Bromley, Kent, BR1 3 bedroom s...   15/05/2023   
3  Gladwell Road, Bromley. 2 bedroom terraced hou...   01/05/2023   
4  Liddon Road, Bromley, Kent, BR1. 1 bedroom apa...   01/05/2023   

                                        Agent Listing_Source   
0                    Edward Ashdale - Bromley   OnTheMartket  \
1          Sinclair Hammelton - Bromley Sales   OnTheMartket   
2          Sinclair Hammelton - Bromley Sales   OnTheMartket   
3  Kinleigh Folkard & Hayward - Bromley Sales   OnTheMartket   
4          Sinclair Hammelton - Bromley Sales   OnTheMartket   

                                         listing_URL  
0  https://www.onthemarket.com/for-sale/property/...  
1  https://www.onthemarket.com/for-sale/property/...  
2  https://www.onthemarket.com/for-sale/property/...  
3  https://www.onthemarket.com/for-sale/property/...  
4  https://www.onthemarket.com/for-sale/property/...

### ii) Cleaning the Bathrooms column

In [285]:
# Strip ".0" from float values while preserving "Not Listed"
om['Bathrooms'] = om['Bathrooms'].apply(lambda x: str(x).rstrip('.0') if x != 'Not Listed' else x)


In [286]:
om

Unique_Id Location Tansaction_Type   
0       BR1S00001OM      BR1           Sales  \
1       BR1S00001OM      BR1           Sales   
2       BR1S00001OM      BR1           Sales   
3       BR1S00001OM      BR1           Sales   
4       BR1S00001OM      BR1           Sales   
...             ...      ...             ...   
14785  WC2RS00008OM     WC2R           Sales   
14786  WC2RS00009OM     WC2R           Sales   
14787  WC2RS00010OM     WC2R           Sales   
14788  WC2RS00011OM     WC2R           Sales   
14789  WC2RS00012OM     WC2R           Sales   

                                Property_Type   
0           3 bedroom terraced house for sale  \
1           2 bedroom terraced house for sale   
2      3 bedroom semi-detached house for sale   
3           2 bedroom terraced house for sale   
4                1 bedroom apartment for sale   
...                                       ...   
14785                 1 bedroom flat for sale   
14786            3 bedroom apartment for sale   
14787                 1 bedroom flat for sale   
14788            1 bedroom apartment for sale   
14789            1 bedroom apartment for sale   

                                                 Address Bedrooms Bathrooms   
0                            Parkside Avenue Bromley BR1        3         1  \
1                       Foxbury Road, Bromley, Kent, BR1        2         1   
2                     Bishops Avenue, Bromley, Kent, BR1        3         1   
3                                 Gladwell Road, Bromley        2         1   
4                        Liddon Road, Bromley, Kent, BR1        1         1   
...                                                  ...      ...       ...   
14785  Clement House, 190 Strand, Covent Garden, Lond...        1         1   
14786            Gladstone House, 190 Strand London WC2R        3         3   
14787                               190 Strand, WC2R 1AB        1         1   
14788        The Piazza Residences, Bull Inn Court, WC2R        1         1   
14789                      Savoy House, 190 Strand, WC2R        1         1   

            Price                                        Description   
0        £550,000  Parkside Avenue Bromley BR1 3 bedroom terraced...  \
1        £525,000  Foxbury Road, Bromley, Kent, BR1 2 bedroom ter...   
2        £575,000  Bishops Avenue, Bromley, Kent, BR1 3 bedroom s...   
3        £475,000  Gladwell Road, Bromley. 2 bedroom terraced hou...   
4        £250,000  Liddon Road, Bromley, Kent, BR1. 1 bedroom apa...   
...           ...                                                ...   
14785  £2,750,000  Clement House, 190 Strand, Covent Garden, Lond...   
14786  £2,195,000  Gladstone House, 190 Strand London WC2R. 3 bed...   
14787  £2,600,000      190 Strand, WC2R 1AB. 1 bedroom flat for sale   
14788    £850,000  The Piazza Residences, Bull Inn Court, WC2R. 1...   
14789  £1,500,000  Savoy House, 190 Strand, WC2R. 1 bedroom apart...   

      Listing_Date                                       Agent Listing_Source   
0       01/05/2023                    Edward Ashdale - Bromley   OnTheMartket  \
1       15/05/2023          Sinclair Hammelton - Bromley Sales   OnTheMartket   
2       15/05/2023          Sinclair Hammelton - Bromley Sales   OnTheMartket   
3       01/05/2023  Kinleigh Folkard & Hayward - Bromley Sales   OnTheMartket   
4       01/05/2023          Sinclair Hammelton - Bromley Sales   OnTheMartket   
...            ...                                         ...            ...   
14785   03/05/2023                                  Not listed   OnTheMartket   
14786   03/05/2023                                  Not listed   OnTheMartket   
14787   03/05/2023                                  Not listed   OnTheMartket   
14788   03/05/2023                                  Not listed   OnTheMartket   
14789   03/05/2023                                  Not listed   OnTheMartket   

                                             listing_URL 

### iii) Cleaning the Bedrooms column

In [287]:
# Checking the values counts
om.Bedrooms.value_counts()

Bedrooms
2             5188
3             3510
1             2843
4             1738
5              713
6              293
Studio         272
Not listed     107
7               72
8               27
9               10
10              10
15               3
11               2
17               1
12               1
Name: count, dtype: int64

In [288]:
# Replacing the studios with not listed
om['Bedrooms'] = om['Bedrooms'].replace('Studio', 'Not listed')

In [289]:
# Checking the values counts
om.Bedrooms.value_counts()

Bedrooms
2             5188
3             3510
1             2843
4             1738
5              713
Not listed     379
6              293
7               72
8               27
9               10
10              10
15               3
11               2
17               1
12               1
Name: count, dtype: int64

Perfect!

# 3. Zoopla

Now we will clean the sales data for Zoopla

## a) Data Accessment

In [290]:
zp.head()

Unique_Id Location Tansaction_Type                      Property_Type   
0  BR1S00001ZP      BR1           Sales  Bishops Avenue, Bromley, Kent BR1  \
1  BR1S00002ZP      BR1           Sales          College Road, Bromley BR1   
2  BR1S00003ZP      BR1           Sales  Marian Gardens, Bromley, Kent BR1   
3  BR1S00004ZP      BR1           Sales    Forde Avenue, Bromley, Kent BR1   
4  BR1S00005ZP      BR1           Sales  Ellen Close, Bickley, Bromley BR1   

                              Address  Bedrooms  Bathrooms       Price   
0  6 bed semi-detached house for sale         6          2  £1,250,000  \
1                 1 bed flat for sale         1          1    £255,000   
2                 2 bed flat for sale         2          1    £550,000   
3       3 bed terraced house for sale         3          1    £600,000   
4  3 bed semi-detached house for sale         3          2    £525,000   

                                         Description             Listing_Date   
0  Chain free and newly rennovated to an exceptio...  Listed on 18th May 2023  \
1  Escape the city hustle with this stunning one-...  Listed on 18th May 2023   
2  Located in the prestigious Hampton Grange deve...  Listed on 18th May 2023   
3  A very generously proportioned mid terraced th...  Listed on 18th May 2023   
4  *Guide Price £525,000 - £550,000* A spacious t...  Listed on 18th May 2023   

                       Agent Listing_Source   
0             Lawson S Stone         Zoopla  \
1          CKB Estate Agents         Zoopla   
2   Key Property Consultants         Zoopla   
3  Daniels Property Services         Zoopla   
4  Daniels Property Services         Zoopla   

                                         listing_URL  
0  https://www.zoopla.co.uk/for-sale/property/br1...  
1  https://www.zoopla.co.uk/for-sale/property/br1...  
2  https://www.zoopla.co.uk/for-sale/property/br1...  
3  https://www.zoopla.co.uk/for-sale/property/br1...  
4  https://www.zoopla.co.uk/for-sale/property/br1...

In [291]:
# Checking the shape of the Dataset
zp.shape

(49494, 13)

In [292]:
# Checking the data types
zp.dtypes

Unique_Id          object
Location           object
Tansaction_Type    object
Property_Type      object
Address            object
Bedrooms            int64
Bathrooms           int64
Price              object
Description        object
Listing_Date       object
Agent              object
Listing_Source     object
listing_URL        object
dtype: object

In [293]:
# Cheching the distribution of values
zp.describe()

Bedrooms     Bathrooms
count  49494.000000  49494.000000
mean       2.441367      1.305653
std        1.316987      0.675686
min        1.000000      1.000000
25%        2.000000      1.000000
50%        2.000000      1.000000
75%        3.000000      1.000000
max       33.000000     18.000000

## b) Dealing with nulls

In [294]:
zp.isnull().sum()

Unique_Id          0
Location           0
Tansaction_Type    0
Property_Type      0
Address            0
Bedrooms           0
Bathrooms          0
Price              0
Description        0
Listing_Date       0
Agent              0
Listing_Source     0
listing_URL        0
dtype: int64

There are no nulls in the dataset. Perfect!!!

## c) Data cleaning

In [295]:
zp.head()

Unique_Id Location Tansaction_Type                      Property_Type   
0  BR1S00001ZP      BR1           Sales  Bishops Avenue, Bromley, Kent BR1  \
1  BR1S00002ZP      BR1           Sales          College Road, Bromley BR1   
2  BR1S00003ZP      BR1           Sales  Marian Gardens, Bromley, Kent BR1   
3  BR1S00004ZP      BR1           Sales    Forde Avenue, Bromley, Kent BR1   
4  BR1S00005ZP      BR1           Sales  Ellen Close, Bickley, Bromley BR1   

                              Address  Bedrooms  Bathrooms       Price   
0  6 bed semi-detached house for sale         6          2  £1,250,000  \
1                 1 bed flat for sale         1          1    £255,000   
2                 2 bed flat for sale         2          1    £550,000   
3       3 bed terraced house for sale         3          1    £600,000   
4  3 bed semi-detached house for sale         3          2    £525,000   

                                         Description             Listing_Date   
0  Chain free and newly rennovated to an exceptio...  Listed on 18th May 2023  \
1  Escape the city hustle with this stunning one-...  Listed on 18th May 2023   
2  Located in the prestigious Hampton Grange deve...  Listed on 18th May 2023   
3  A very generously proportioned mid terraced th...  Listed on 18th May 2023   
4  *Guide Price £525,000 - £550,000* A spacious t...  Listed on 18th May 2023   

                       Agent Listing_Source   
0             Lawson S Stone         Zoopla  \
1          CKB Estate Agents         Zoopla   
2   Key Property Consultants         Zoopla   
3  Daniels Property Services         Zoopla   
4  Daniels Property Services         Zoopla   

                                         listing_URL  
0  https://www.zoopla.co.uk/for-sale/property/br1...  
1  https://www.zoopla.co.uk/for-sale/property/br1...  
2  https://www.zoopla.co.uk/for-sale/property/br1...  
3  https://www.zoopla.co.uk/for-sale/property/br1...  
4  https://www.zoopla.co.uk/for-sale/property/br1...

### i) Cleaning the Listing_Date column

We will attempt to clean the Listing_Date to remove the 'Listed on' string attached to the listed date

In [296]:
# Strip "Listed on" from Date values while preserving 
zp['Listing_Date'] = zp['Listing_Date'].str.replace('Listed on', '').str.strip()

In [297]:
zp.head()

Unique_Id Location Tansaction_Type                      Property_Type   
0  BR1S00001ZP      BR1           Sales  Bishops Avenue, Bromley, Kent BR1  \
1  BR1S00002ZP      BR1           Sales          College Road, Bromley BR1   
2  BR1S00003ZP      BR1           Sales  Marian Gardens, Bromley, Kent BR1   
3  BR1S00004ZP      BR1           Sales    Forde Avenue, Bromley, Kent BR1   
4  BR1S00005ZP      BR1           Sales  Ellen Close, Bickley, Bromley BR1   

                              Address  Bedrooms  Bathrooms       Price   
0  6 bed semi-detached house for sale         6          2  £1,250,000  \
1                 1 bed flat for sale         1          1    £255,000   
2                 2 bed flat for sale         2          1    £550,000   
3       3 bed terraced house for sale         3          1    £600,000   
4  3 bed semi-detached house for sale         3          2    £525,000   

                                         Description   Listing_Date   
0  Chain free and newly rennovated to an exceptio...  18th May 2023  \
1  Escape the city hustle with this stunning one-...  18th May 2023   
2  Located in the prestigious Hampton Grange deve...  18th May 2023   
3  A very generously proportioned mid terraced th...  18th May 2023   
4  *Guide Price £525,000 - £550,000* A spacious t...  18th May 2023   

                       Agent Listing_Source   
0             Lawson S Stone         Zoopla  \
1          CKB Estate Agents         Zoopla   
2   Key Property Consultants         Zoopla   
3  Daniels Property Services         Zoopla   
4  Daniels Property Services         Zoopla   

                                         listing_URL  
0  https://www.zoopla.co.uk/for-sale/property/br1...  
1  https://www.zoopla.co.uk/for-sale/property/br1...  
2  https://www.zoopla.co.uk/for-sale/property/br1...  
3  https://www.zoopla.co.uk/for-sale/property/br1...  
4  https://www.zoopla.co.uk/for-sale/property/br1...

Now, we will convert the Listing dates into the format dd/mm/yyyy, for uniformity with the other datasets

In [298]:
# Convert the date format to 'dd/mm/yyyy'
zp['Listing_Date'] = zp['Listing_Date'].str.replace('th', '').str.replace('st', '').str.replace('nd', '').str.replace('rd', '').str.replace(' ', '/')

# Convert the month name to the month number
try: 
    zp['Listing_Date'] = pd.to_datetime(zp['Listing_Date'], format='%d/%B/%Y').dt.strftime('%d/%m/%Y')
except ValueError:
    pass

# Convert the month name to the month number
try: 
    # Update the month name to month number conversion
    zp['Listing_Date'] = zp['Listing_Date'].str.replace('/Jan/', '/01/').str.replace('/Feb/', '/02/').str.replace('/Mar/', '/03/').str.replace('/Apr/', '/04/').str.replace('/May/', '/05/').str.replace('/Jun/', '/06/').str.replace('/Jul/', '/07/').str.replace('/Aug/', '/08/').str.replace('/Sep/', '/09/').str.replace('/Oct/', '/10/').str.replace('/Nov/', '/11/').str.replace('/Dec/', '/12/')
except ValueError:
    pass


Convert any non-date values in the Listed_Date column to 'Not Listed'

In [299]:
# Convert the column to string type
zp['Listing_Date'] = zp['Listing_Date'].astype(str)

# Check if the value matches the date format
mask = zp['Listing_Date'].str.match(r'\d{2}/\d{2}/\d{4}')

# Replace non-matching values with 'Not Listed'
zp.loc[~mask, 'Listing_Date'] = 'Not Listed'

In [300]:
zp.head()

Unique_Id Location Tansaction_Type                      Property_Type   
0  BR1S00001ZP      BR1           Sales  Bishops Avenue, Bromley, Kent BR1  \
1  BR1S00002ZP      BR1           Sales          College Road, Bromley BR1   
2  BR1S00003ZP      BR1           Sales  Marian Gardens, Bromley, Kent BR1   
3  BR1S00004ZP      BR1           Sales    Forde Avenue, Bromley, Kent BR1   
4  BR1S00005ZP      BR1           Sales  Ellen Close, Bickley, Bromley BR1   

                              Address  Bedrooms  Bathrooms       Price   
0  6 bed semi-detached house for sale         6          2  £1,250,000  \
1                 1 bed flat for sale         1          1    £255,000   
2                 2 bed flat for sale         2          1    £550,000   
3       3 bed terraced house for sale         3          1    £600,000   
4  3 bed semi-detached house for sale         3          2    £525,000   

                                         Description Listing_Date   
0  Chain free and newly rennovated to an exceptio...   18/05/2023  \
1  Escape the city hustle with this stunning one-...   18/05/2023   
2  Located in the prestigious Hampton Grange deve...   18/05/2023   
3  A very generously proportioned mid terraced th...   18/05/2023   
4  *Guide Price £525,000 - £550,000* A spacious t...   18/05/2023   

                       Agent Listing_Source   
0             Lawson S Stone         Zoopla  \
1          CKB Estate Agents         Zoopla   
2   Key Property Consultants         Zoopla   
3  Daniels Property Services         Zoopla   
4  Daniels Property Services         Zoopla   

                                         listing_URL  
0  https://www.zoopla.co.uk/for-sale/property/br1...  
1  https://www.zoopla.co.uk/for-sale/property/br1...  
2  https://www.zoopla.co.uk/for-sale/property/br1...  
3  https://www.zoopla.co.uk/for-sale/property/br1...  
4  https://www.zoopla.co.uk/for-sale/property/br1...

Looks great!

<h1 style='background-color: Indigo; padding: 10px; color: white'> Rent data </h1>

We will import the rent data for the three websites

In [301]:
# Rightmove
rm1 = pd.read_csv('rightmove_rent_data.csv')

# On the Market
om1 = pd.read_csv('OnTheMarket_rent_data.csv')

#Zoopla
zp1 = pd.read_csv('Zoopla_rent-data.csv')

We will remove the unwanted column 'Unnamed from all the datasets'

In [302]:
# Rightmove
rm1 = rm1.drop('Unnamed: 0', axis=1)

# On the Market
om1 = om1.drop('Unnamed: 0', axis=1)

#Zoopla
zp1 = zp1.drop('Unnamed: 0', axis=1)

# Rightmove

Now we will assess and clean the rent data for rightmove

## a) Data Assesment

In [303]:
rm1.head()

Unique_Id Location Tansaction_Type   Property_Type   
0  BR1R00001      BR1            Rent        Terraced  \
1  BR1R00002      BR1            Rent  End of Terrace   
2  BR1R00003      BR1            Rent            Flat   
3  BR1R00004      BR1            Rent          Studio   
4  BR1R00005      BR1            Rent            Flat   

                                             Address  Bedrooms  Bathrooms   
0                       Hillcrest Road, Bromley, BR1         3          1  \
1                         Pendragon Road Bromley BR1         5          2   
2                             Hillside, Bickley, BR1         2          1   
3  **ALL BILLS INCLUDED** Coniston Road, Bromley,...         1          1   
4                           Dale Court, Bromley, BR1         3          1   

        Price                                        Description   
0  £2,000 pcm  We are delighted to present to the market this...  \
1  £2,500 pcm  We are delighted to present this wonderfully s...   
2  £1,500 pcm  *ZERO DEPOSIT OPTION AVAILABLE* *AVAILABLE NOW...   
3  £1,200 pcm  **ALL BILLS INCLUDED** - BEAUTIFUL, LARGE, MOD...   
4  £1,600 pcm  We are pleased to market this 3 bedroom, 1 bat...   

            Listing_Date                                              Agent   
0  Reduced on 04/05/2023  Reduced on 04/05/2023 by Daniels Property Serv...  \
1          Reduced today                 Reduced today by Acorn, Grove Park   
2          Reduced today   Reduced today by Andrew Reeves, Bromley-Lettings   
3          Reduced today   Reduced today by Andrew Reeves, Bromley-Lettings   
4    Added on 07/05/2023            Added on 07/05/2023 by OpenRent, London   

  Listing_Source                                        listing_URL  
0      RightMove  https://www.rightmove.co.uk/property-to-rent/f...  
1      RightMove  https://www.rightmove.co.uk/property-to-rent/f...  
2      RightMove  https://www.rightmove.co.uk/property-to-rent/f...  
3      RightMove  https://www.rightmove.co.uk/property-to-rent/f...  
4      RightMove  https://www.rightmove.co.uk/property-to-rent/f...

In [304]:
# Checking the shape of the data
rm1.shape

(21126, 13)

In [305]:
# Checking the distribution of values
rm1.describe()

Bedrooms     Bathrooms
count  21126.000000  21126.000000
mean       2.147259      1.712203
std        1.171554      0.935005
min        1.000000      1.000000
25%        1.000000      1.000000
50%        2.000000      2.000000
75%        3.000000      2.000000
max       20.000000     12.000000

In [306]:
# Assessing the data types
rm1.dtypes

Unique_Id          object
Location           object
Tansaction_Type    object
Property_Type      object
Address            object
Bedrooms            int64
Bathrooms           int64
Price              object
Description        object
Listing_Date       object
Agent              object
Listing_Source     object
listing_URL        object
dtype: object

## b) Dealing with nulls

In [307]:
rm1.isnull().sum()

Unique_Id          0
Location           0
Tansaction_Type    0
Property_Type      0
Address            0
Bedrooms           0
Bathrooms          0
Price              0
Description        0
Listing_Date       7
Agent              0
Listing_Source     0
listing_URL        0
dtype: int64

Only the Listing_Date column has nulls

### i) Dealing with nulls in the Listing_Date column

We will replace the nulls with 'Not Listed'

In [308]:
rm1['Listing_Date'].fillna('Not listed', inplace=True)

In [309]:
rm1.isnull().sum()

Unique_Id          0
Location           0
Tansaction_Type    0
Property_Type      0
Address            0
Bedrooms           0
Bathrooms          0
Price              0
Description        0
Listing_Date       0
Agent              0
Listing_Source     0
listing_URL        0
dtype: int64

Nulls succesfully dealth with

## c) Data Cleaning

In [310]:
rm1.head()

Unique_Id Location Tansaction_Type   Property_Type   
0  BR1R00001      BR1            Rent        Terraced  \
1  BR1R00002      BR1            Rent  End of Terrace   
2  BR1R00003      BR1            Rent            Flat   
3  BR1R00004      BR1            Rent          Studio   
4  BR1R00005      BR1            Rent            Flat   

                                             Address  Bedrooms  Bathrooms   
0                       Hillcrest Road, Bromley, BR1         3          1  \
1                         Pendragon Road Bromley BR1         5          2   
2                             Hillside, Bickley, BR1         2          1   
3  **ALL BILLS INCLUDED** Coniston Road, Bromley,...         1          1   
4                           Dale Court, Bromley, BR1         3          1   

        Price                                        Description   
0  £2,000 pcm  We are delighted to present to the market this...  \
1  £2,500 pcm  We are delighted to present this wonderfully s...   
2  £1,500 pcm  *ZERO DEPOSIT OPTION AVAILABLE* *AVAILABLE NOW...   
3  £1,200 pcm  **ALL BILLS INCLUDED** - BEAUTIFUL, LARGE, MOD...   
4  £1,600 pcm  We are pleased to market this 3 bedroom, 1 bat...   

            Listing_Date                                              Agent   
0  Reduced on 04/05/2023  Reduced on 04/05/2023 by Daniels Property Serv...  \
1          Reduced today                 Reduced today by Acorn, Grove Park   
2          Reduced today   Reduced today by Andrew Reeves, Bromley-Lettings   
3          Reduced today   Reduced today by Andrew Reeves, Bromley-Lettings   
4    Added on 07/05/2023            Added on 07/05/2023 by OpenRent, London   

  Listing_Source                                        listing_URL  
0      RightMove  https://www.rightmove.co.uk/property-to-rent/f...  
1      RightMove  https://www.rightmove.co.uk/property-to-rent/f...  
2      RightMove  https://www.rightmove.co.uk/property-to-rent/f...  
3      RightMove  https://www.rightmove.co.uk/property-to-rent/f...  
4      RightMove  https://www.rightmove.co.uk/property-to-rent/f...

Now, we will modify the Listed date column

In [311]:
# Removing the 'Added on' from the values
rm1['Listing_Date'] = rm1['Listing_Date'].str.replace('Added on ', '', regex=True)
rm1['Listing_Date'] = rm1['Listing_Date'].str.replace('Reduced on ', '', regex=True)

Convert any non-date values in the Listed_Date column to 'Not Listed'

In [312]:
# Convert the column to string type
rm1['Listing_Date'] = rm1['Listing_Date'].astype(str)

# Check if the value matches the date format
mask = rm1['Listing_Date'].str.match(r'\d{2}/\d{2}/\d{4}')

# Replace non-matching values with 'Not Listed'
rm1.loc[~mask, 'Listing_Date'] = 'Not Listed'

In [313]:
rm1.head()

Unique_Id Location Tansaction_Type   Property_Type   
0  BR1R00001      BR1            Rent        Terraced  \
1  BR1R00002      BR1            Rent  End of Terrace   
2  BR1R00003      BR1            Rent            Flat   
3  BR1R00004      BR1            Rent          Studio   
4  BR1R00005      BR1            Rent            Flat   

                                             Address  Bedrooms  Bathrooms   
0                       Hillcrest Road, Bromley, BR1         3          1  \
1                         Pendragon Road Bromley BR1         5          2   
2                             Hillside, Bickley, BR1         2          1   
3  **ALL BILLS INCLUDED** Coniston Road, Bromley,...         1          1   
4                           Dale Court, Bromley, BR1         3          1   

        Price                                        Description Listing_Date   
0  £2,000 pcm  We are delighted to present to the market this...   04/05/2023  \
1  £2,500 pcm  We are delighted to present this wonderfully s...   Not Listed   
2  £1,500 pcm  *ZERO DEPOSIT OPTION AVAILABLE* *AVAILABLE NOW...   Not Listed   
3  £1,200 pcm  **ALL BILLS INCLUDED** - BEAUTIFUL, LARGE, MOD...   Not Listed   
4  £1,600 pcm  We are pleased to market this 3 bedroom, 1 bat...   07/05/2023   

                                               Agent Listing_Source   
0  Reduced on 04/05/2023 by Daniels Property Serv...      RightMove  \
1                 Reduced today by Acorn, Grove Park      RightMove   
2   Reduced today by Andrew Reeves, Bromley-Lettings      RightMove   
3   Reduced today by Andrew Reeves, Bromley-Lettings      RightMove   
4            Added on 07/05/2023 by OpenRent, London      RightMove   

                                         listing_URL  
0  https://www.rightmove.co.uk/property-to-rent/f...  
1  https://www.rightmove.co.uk/property-to-rent/f...  
2  https://www.rightmove.co.uk/property-to-rent/f...  
3  https://www.rightmove.co.uk/property-to-rent/f...  
4  https://www.rightmove.co.uk/property-to-rent/f...

### ii) Cleaning the Agent column

Modifying the Agent column

In [314]:
# Removing all strings before the word 'by'
try:
    rm1['Agent'] = rm1['Agent'].str.split('by', n=1).str[1].str.strip()
except ValueError:
    pass

In [315]:
# Replace all nulls with 'Not listed'
rm1['Agent'].fillna('Not listed', inplace=True)

In [316]:
rm1.head()

Unique_Id Location Tansaction_Type   Property_Type   
0  BR1R00001      BR1            Rent        Terraced  \
1  BR1R00002      BR1            Rent  End of Terrace   
2  BR1R00003      BR1            Rent            Flat   
3  BR1R00004      BR1            Rent          Studio   
4  BR1R00005      BR1            Rent            Flat   

                                             Address  Bedrooms  Bathrooms   
0                       Hillcrest Road, Bromley, BR1         3          1  \
1                         Pendragon Road Bromley BR1         5          2   
2                             Hillside, Bickley, BR1         2          1   
3  **ALL BILLS INCLUDED** Coniston Road, Bromley,...         1          1   
4                           Dale Court, Bromley, BR1         3          1   

        Price                                        Description Listing_Date   
0  £2,000 pcm  We are delighted to present to the market this...   04/05/2023  \
1  £2,500 pcm  We are delighted to present this wonderfully s...   Not Listed   
2  £1,500 pcm  *ZERO DEPOSIT OPTION AVAILABLE* *AVAILABLE NOW...   Not Listed   
3  £1,200 pcm  **ALL BILLS INCLUDED** - BEAUTIFUL, LARGE, MOD...   Not Listed   
4  £1,600 pcm  We are pleased to market this 3 bedroom, 1 bat...   07/05/2023   

                                Agent Listing_Source   
0  Daniels Property Services, Bromley      RightMove  \
1                   Acorn, Grove Park      RightMove   
2     Andrew Reeves, Bromley-Lettings      RightMove   
3     Andrew Reeves, Bromley-Lettings      RightMove   
4                    OpenRent, London      RightMove   

                                         listing_URL  
0  https://www.rightmove.co.uk/property-to-rent/f...  
1  https://www.rightmove.co.uk/property-to-rent/f...  
2  https://www.rightmove.co.uk/property-to-rent/f...  
3  https://www.rightmove.co.uk/property-to-rent/f...  
4  https://www.rightmove.co.uk/property-to-rent/f...

Looks great!!!

# On the Market

Now we will clean the rent data for On the Market

## a) Data Assessment

In [317]:
om1.head()

Unique_Id Location Tansaction_Type                Property_Type   
0  BR1S00001OM      BR1            Rent  2 bedroom apartment to rent  \
1  BR1S00002OM      BR1            Rent  2 bedroom apartment to rent   
2  BR1S00003OM      BR1            Rent      5 bedroom house to rent   
3  BR1S00004OM      BR1            Rent               Studio to rent   
4  BR1S00005OM      BR1            Rent       2 bedroom flat to rent   

                                         Address Bedrooms Bathrooms   
0                          Park Hill Bickley BR1        2         1  \
1                       237 Downham Way, Bromley        2         1   
2  Edgeborough Way, Sundridge Park, Bromley, BR1        5         2   
3                       Bromley Hill, London BR1   Studio         1   
4                     Widmore Road, Bromley, BR1        2         2   

                  Price                                        Description   
0  £1,700 pcm (£392 pw)  Park Hill Bickley BR1. 2 bedroom apartment to ...  \
1  £1,650 pcm (£381 pw)  237 Downham Way, Bromley. 2 bedroom apartment ...   
2  £2,800 pcm (£646 pw)  Edgeborough Way, Sundridge Park, Bromley, BR1....   
3  £1,050 pcm (£242 pw)           Bromley Hill, London BR1. Studio to rent   
4  £1,850 pcm (£427 pw)  Widmore Road, Bromley, BR1. 2 bedroom flat to ...   

           Listing_Date                                              Agent   
0  OnTheMarket < 7 days  Kinleigh Folkard & Hayward - Chislehurst Lettings  \
1  OnTheMarket < 7 days     Simon & Simon Property Services - Hampton Wick   
2  OnTheMarket < 7 days                                  Foxtons - Bromley   
3      Reduced < 7 days                             Milegate - Cricklewood   
4  OnTheMarket < 7 days                                Leo Newman - London   

  Listing_Source                                        listing_URL   
0   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/  \
1   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/   
2   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/   
3   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/   
4   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/   

  Current_date  
0   18/05/2023  
1   18/05/2023  
2   18/05/2023  
3   18/05/2023  
4   18/05/2023

In [318]:
# Checking the shape of the Dataset
om1.shape

(22516, 14)

In [319]:
# Checking the datatype
om1.dtypes

Unique_Id          object
Location           object
Tansaction_Type    object
Property_Type      object
Address            object
Bedrooms           object
Bathrooms          object
Price              object
Description        object
Listing_Date       object
Agent              object
Listing_Source     object
listing_URL        object
Current_date       object
dtype: object

## b) Dealing with nulls

In [320]:
om1.isnull().sum()

Unique_Id             0
Location              0
Tansaction_Type       0
Property_Type         0
Address               0
Bedrooms            647
Bathrooms          1806
Price                 3
Description           0
Listing_Date          0
Agent               103
Listing_Source        0
listing_URL           0
Current_date          0
dtype: int64

Now, we will replace all the nulls with 'Not listed'

In [321]:
# Replacing nulls in the Bedroom column with 'Not Listed'
om1['Bedrooms'].fillna('Not listed', inplace=True)

# Replacing nulls in the Bathrooms column with 'Not Listed'
om1['Bathrooms'].fillna('Not listed', inplace=True)

# Replacing nulls in the Agent column with 'Not Listed'
om1['Agent'].fillna('Not listed', inplace=True)

# Replacing nulls in the Price column with 'Not Listed'
om1['Price'].fillna('Not listed', inplace=True)

In [322]:
om1.isnull().sum()

Unique_Id          0
Location           0
Tansaction_Type    0
Property_Type      0
Address            0
Bedrooms           0
Bathrooms          0
Price              0
Description        0
Listing_Date       0
Agent              0
Listing_Source     0
listing_URL        0
Current_date       0
dtype: int64

Great!!!

## c) Data Cleaning

### i) Cleaning the Listing_Date Column

Here are the values available in the Listing_Date

In [323]:
om1.Listing_Date.value_counts()

Listing_Date
OnTheMarket > 14 days    14253
OnTheMarket < 7 days      3439
OnTheMarket < 14 days     2496
Reduced < 7 days           701
Reduced < 14 days          583
OnTheMarket today          439
OnTheMarket yesterday      385
Reduced today              112
Reduced yesterday           75
Reduced > 14 days           32
Listing_Date                 1
Name: count, dtype: int64

We will repeat the same process as we did with the On the market sales data

In [324]:
# Convert 'Current_date' column to datetime type
om1['Current_date'] = pd.to_datetime(om1['Current_date'], format='%d/%m/%Y')

# Calculate the new 'Date_main' column based on different conditions
om1['Date_main'] = om1.apply(
    lambda row: (
        (row['Current_date']).strftime('%d/%m/%Y')
        if row['Listing_Date'] in ['Reduced today', 'OnTheMarket today']
        else (row['Current_date'] - timedelta(days=1)).strftime('%d/%m/%Y')
        if row['Listing_Date'] in ['Reduced yesterday', 'OnTheMarket yesterday']
        else (row['Current_date'] - timedelta(days=14)).strftime('%d/%m/%Y')  # > (current date - 14 days)
        if row['Listing_Date'] in ['OnTheMarket > 14 days', 'Reduced > 14 days']
        else (row['Current_date'] - timedelta(days=14)).strftime('%d/%m/%Y')  # < (current date - 14 days)
        if row['Listing_Date'] in ['OnTheMarket < 14 days', 'Reduced < 14 days']
        else (row['Current_date'] - timedelta(days=7)).strftime('%d/%m/%Y')  # < (current date - 7 days)
        if row['Listing_Date'] in ['OnTheMarket < 7 days', 'Reduced < 7 days']
        else None  # Add any other conditions or default value if needed
    ),
    axis=1
)

# Convert the column to datetime type
om1['Date_main'] = pd.to_datetime(om1['Date_main'], format='%d/%m/%Y')

# Convert the values to the desired format
om1['Date_main'] = om1['Date_main'].dt.strftime('%d/%m/%Y')


In [325]:
om1.head()

Unique_Id Location Tansaction_Type                Property_Type   
0  BR1S00001OM      BR1            Rent  2 bedroom apartment to rent  \
1  BR1S00002OM      BR1            Rent  2 bedroom apartment to rent   
2  BR1S00003OM      BR1            Rent      5 bedroom house to rent   
3  BR1S00004OM      BR1            Rent               Studio to rent   
4  BR1S00005OM      BR1            Rent       2 bedroom flat to rent   

                                         Address Bedrooms Bathrooms   
0                          Park Hill Bickley BR1        2         1  \
1                       237 Downham Way, Bromley        2         1   
2  Edgeborough Way, Sundridge Park, Bromley, BR1        5         2   
3                       Bromley Hill, London BR1   Studio         1   
4                     Widmore Road, Bromley, BR1        2         2   

                  Price                                        Description   
0  £1,700 pcm (£392 pw)  Park Hill Bickley BR1. 2 bedroom apartment to ...  \
1  £1,650 pcm (£381 pw)  237 Downham Way, Bromley. 2 bedroom apartment ...   
2  £2,800 pcm (£646 pw)  Edgeborough Way, Sundridge Park, Bromley, BR1....   
3  £1,050 pcm (£242 pw)           Bromley Hill, London BR1. Studio to rent   
4  £1,850 pcm (£427 pw)  Widmore Road, Bromley, BR1. 2 bedroom flat to ...   

           Listing_Date                                              Agent   
0  OnTheMarket < 7 days  Kinleigh Folkard & Hayward - Chislehurst Lettings  \
1  OnTheMarket < 7 days     Simon & Simon Property Services - Hampton Wick   
2  OnTheMarket < 7 days                                  Foxtons - Bromley   
3      Reduced < 7 days                             Milegate - Cricklewood   
4  OnTheMarket < 7 days                                Leo Newman - London   

  Listing_Source                                        listing_URL   
0   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/  \
1   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/   
2   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/   
3   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/   
4   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/   

  Current_date   Date_main  
0   2023-05-18  11/05/2023  
1   2023-05-18  11/05/2023  
2   2023-05-18  11/05/2023  
3   2023-05-18  11/05/2023  
4   2023-05-18  11/05/2023

In [326]:
om1.shape

(22516, 15)

In [327]:
# Drop the 'Current_date' and 'Listing_Date' columns
om1.drop(['Current_date', 'Listing_Date'], axis=1, inplace=True)

For uniformity with the datsets from other websites, the Date_main column will be renamed as Listing_Date

In [328]:
# Rename the 'Date_main' column to 'Listing_Date'
om1.rename(columns={'Date_main': 'Listing_Date'}, inplace=True)

In [329]:
om1.head()

Unique_Id Location Tansaction_Type                Property_Type   
0  BR1S00001OM      BR1            Rent  2 bedroom apartment to rent  \
1  BR1S00002OM      BR1            Rent  2 bedroom apartment to rent   
2  BR1S00003OM      BR1            Rent      5 bedroom house to rent   
3  BR1S00004OM      BR1            Rent               Studio to rent   
4  BR1S00005OM      BR1            Rent       2 bedroom flat to rent   

                                         Address Bedrooms Bathrooms   
0                          Park Hill Bickley BR1        2         1  \
1                       237 Downham Way, Bromley        2         1   
2  Edgeborough Way, Sundridge Park, Bromley, BR1        5         2   
3                       Bromley Hill, London BR1   Studio         1   
4                     Widmore Road, Bromley, BR1        2         2   

                  Price                                        Description   
0  £1,700 pcm (£392 pw)  Park Hill Bickley BR1. 2 bedroom apartment to ...  \
1  £1,650 pcm (£381 pw)  237 Downham Way, Bromley. 2 bedroom apartment ...   
2  £2,800 pcm (£646 pw)  Edgeborough Way, Sundridge Park, Bromley, BR1....   
3  £1,050 pcm (£242 pw)           Bromley Hill, London BR1. Studio to rent   
4  £1,850 pcm (£427 pw)  Widmore Road, Bromley, BR1. 2 bedroom flat to ...   

                                               Agent Listing_Source   
0  Kinleigh Folkard & Hayward - Chislehurst Lettings   OnTheMartket  \
1     Simon & Simon Property Services - Hampton Wick   OnTheMartket   
2                                  Foxtons - Bromley   OnTheMartket   
3                             Milegate - Cricklewood   OnTheMartket   
4                                Leo Newman - London   OnTheMartket   

                                         listing_URL Listing_Date  
0  https://www.onthemarket.com/to-rent/property/br1/   11/05/2023  
1  https://www.onthemarket.com/to-rent/property/br1/   11/05/2023  
2  https://www.onthemarket.com/to-rent/property/br1/   11/05/2023  
3  https://www.onthemarket.com/to-rent/property/br1/   11/05/2023  
4  https://www.onthemarket.com/to-rent/property/br1/   11/05/2023

Now we will move the new listed date column to it's right position

In [330]:
# Get the column names
columns = om1.columns.tolist()

# Specify the position where you want to move the column (e.g., from 4th position to 1st position)
from_position = 12
to_position = 9

# Move the column to the desired position
columns.insert(to_position, columns.pop(from_position))

# Reassign the columns in the DataFrame with the updated order
om1 = om1[columns]



In [331]:
om1.head()

Unique_Id Location Tansaction_Type                Property_Type   
0  BR1S00001OM      BR1            Rent  2 bedroom apartment to rent  \
1  BR1S00002OM      BR1            Rent  2 bedroom apartment to rent   
2  BR1S00003OM      BR1            Rent      5 bedroom house to rent   
3  BR1S00004OM      BR1            Rent               Studio to rent   
4  BR1S00005OM      BR1            Rent       2 bedroom flat to rent   

                                         Address Bedrooms Bathrooms   
0                          Park Hill Bickley BR1        2         1  \
1                       237 Downham Way, Bromley        2         1   
2  Edgeborough Way, Sundridge Park, Bromley, BR1        5         2   
3                       Bromley Hill, London BR1   Studio         1   
4                     Widmore Road, Bromley, BR1        2         2   

                  Price                                        Description   
0  £1,700 pcm (£392 pw)  Park Hill Bickley BR1. 2 bedroom apartment to ...  \
1  £1,650 pcm (£381 pw)  237 Downham Way, Bromley. 2 bedroom apartment ...   
2  £2,800 pcm (£646 pw)  Edgeborough Way, Sundridge Park, Bromley, BR1....   
3  £1,050 pcm (£242 pw)           Bromley Hill, London BR1. Studio to rent   
4  £1,850 pcm (£427 pw)  Widmore Road, Bromley, BR1. 2 bedroom flat to ...   

  Listing_Date                                              Agent   
0   11/05/2023  Kinleigh Folkard & Hayward - Chislehurst Lettings  \
1   11/05/2023     Simon & Simon Property Services - Hampton Wick   
2   11/05/2023                                  Foxtons - Bromley   
3   11/05/2023                             Milegate - Cricklewood   
4   11/05/2023                                Leo Newman - London   

  Listing_Source                                        listing_URL  
0   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/  
1   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/  
2   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/  
3   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/  
4   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/

### ii) Cleaning the Bathrooms column

In [332]:
# Strip ".0" from float values while preserving "Not Listed"
om1['Bathrooms'] = om1['Bathrooms'].apply(lambda x: str(x).rstrip('.0') if x != 'Not Listed' else x)


In [333]:
om1.head()

Unique_Id Location Tansaction_Type                Property_Type   
0  BR1S00001OM      BR1            Rent  2 bedroom apartment to rent  \
1  BR1S00002OM      BR1            Rent  2 bedroom apartment to rent   
2  BR1S00003OM      BR1            Rent      5 bedroom house to rent   
3  BR1S00004OM      BR1            Rent               Studio to rent   
4  BR1S00005OM      BR1            Rent       2 bedroom flat to rent   

                                         Address Bedrooms Bathrooms   
0                          Park Hill Bickley BR1        2         1  \
1                       237 Downham Way, Bromley        2         1   
2  Edgeborough Way, Sundridge Park, Bromley, BR1        5         2   
3                       Bromley Hill, London BR1   Studio         1   
4                     Widmore Road, Bromley, BR1        2         2   

                  Price                                        Description   
0  £1,700 pcm (£392 pw)  Park Hill Bickley BR1. 2 bedroom apartment to ...  \
1  £1,650 pcm (£381 pw)  237 Downham Way, Bromley. 2 bedroom apartment ...   
2  £2,800 pcm (£646 pw)  Edgeborough Way, Sundridge Park, Bromley, BR1....   
3  £1,050 pcm (£242 pw)           Bromley Hill, London BR1. Studio to rent   
4  £1,850 pcm (£427 pw)  Widmore Road, Bromley, BR1. 2 bedroom flat to ...   

  Listing_Date                                              Agent   
0   11/05/2023  Kinleigh Folkard & Hayward - Chislehurst Lettings  \
1   11/05/2023     Simon & Simon Property Services - Hampton Wick   
2   11/05/2023                                  Foxtons - Bromley   
3   11/05/2023                             Milegate - Cricklewood   
4   11/05/2023                                Leo Newman - London   

  Listing_Source                                        listing_URL  
0   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/  
1   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/  
2   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/  
3   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/  
4   OnTheMartket  https://www.onthemarket.com/to-rent/property/br1/

### iii) Cleaning the Bedrooms column

In [334]:
# Checking the values counts
om1.Bedrooms.value_counts()


Bedrooms
2             8014
1             5508
3             3924
Studio        1910
4             1390
5              749
Not listed     647
6              243
7               92
8               24
10               9
11               3
12               1
Bedrooms         1
9                1
Name: count, dtype: int64

In [335]:
# Replacing the studios with not listed
om1['Bedrooms'] = om1['Bedrooms'].replace('Studio', 'Not listed')
om1['Bedrooms'] = om1['Bedrooms'].replace('Bedrooms', 'Not listed')

In [336]:
# Checking the values counts
om1.Bedrooms.value_counts()

Bedrooms
2             8014
1             5508
3             3924
Not listed    2558
4             1390
5              749
6              243
7               92
8               24
10               9
11               3
12               1
9                1
Name: count, dtype: int64

# 3. Zoopla

Now we will clean the sales data for Zoopla

## a) Data Assessment

In [337]:
zp1.head()

Unique_Id Location Tansaction_Type   
0  BR1R00001ZP      BR1            Rent  \
1  BR1R00002ZP      BR1            Rent   
2  BR1R00003ZP      BR1            Rent   
3  BR1R00004ZP      BR1            Rent   
4  BR1R00005ZP      BR1            Rent   

                              Property_Type                   Address   
0                    Park Road, Bromley BR1  1 bed maisonette to rent  \
1                Highland Road, Bromley BR1        3 bed flat to rent   
2                  Downham Way, Bromley BR1        2 bed flat to rent   
3            Downham Way, Bromley, Kent BR1        2 bed flat to rent   
4  Downham Way, Bromley, Greater London BR1        2 bed flat to rent   

   Bedrooms  Bathrooms       Price   
0         1          1  £1,450 pcm  \
1         3          1  £2,250 pcm   
2         2          1  £1,700 pcm   
3         2          1  £1,300 pcm   
4         2          1  £1,250 pcm   

                                         Description   
0  Lawson S Stone are proud to present this amazi...  \
1  **two balconies & parking** inclusive of water...   
2  Two bedroom second floor apartment located on ...   
3  Recently refurbished 2 bedroom flat available ...   
4  Newly refurbished 2 bedroom flat available to ...   

                   Listing_Date                      Agent Listing_Source   
0       Listed on 25th May 2023             Lawson S Stone         Zoopla  \
1   Available from 1st Jun 2023          CKB Estate Agents         Zoopla   
2       Listed on 25th May 2023   Key Property Consultants         Zoopla   
3  Available from 10th Jul 2023  Daniels Property Services         Zoopla   
4       Listed on 25th May 2023  Daniels Property Services         Zoopla   

                                         listing_URL  
0  https://www.zoopla.co.uk/to-rent/property/br1/...  
1  https://www.zoopla.co.uk/to-rent/property/br1/...  
2  https://www.zoopla.co.uk/to-rent/property/br1/...  
3  https://www.zoopla.co.uk/to-rent/property/br1/...  
4  https://www.zoopla.co.uk/to-rent/property/br1/...

In [338]:
# Checking the shape
zp1.shape

(21873, 13)

In [339]:
# Checking the distribution of values
zp1.describe()

Bedrooms     Bathrooms
count  21873.000000  21873.000000
mean       2.105655      1.136744
std        1.161534      0.475439
min        1.000000      1.000000
25%        1.000000      1.000000
50%        2.000000      1.000000
75%        3.000000      1.000000
max       13.000000      9.000000

## b) Dealing with nulls

In [340]:
zp1.isnull().sum()

Unique_Id          0
Location           0
Tansaction_Type    0
Property_Type      0
Address            0
Bedrooms           0
Bathrooms          0
Price              0
Description        1
Listing_Date       0
Agent              0
Listing_Source     0
listing_URL        0
dtype: int64

We will replace the null in the description column with 'Not listed'

In [341]:
zp1['Description'].fillna('Not listed', inplace=True)

In [342]:
zp1.isnull().sum()

Unique_Id          0
Location           0
Tansaction_Type    0
Property_Type      0
Address            0
Bedrooms           0
Bathrooms          0
Price              0
Description        0
Listing_Date       0
Agent              0
Listing_Source     0
listing_URL        0
dtype: int64

## c) Data cleaning

In [343]:
zp1.head()

Unique_Id Location Tansaction_Type   
0  BR1R00001ZP      BR1            Rent  \
1  BR1R00002ZP      BR1            Rent   
2  BR1R00003ZP      BR1            Rent   
3  BR1R00004ZP      BR1            Rent   
4  BR1R00005ZP      BR1            Rent   

                              Property_Type                   Address   
0                    Park Road, Bromley BR1  1 bed maisonette to rent  \
1                Highland Road, Bromley BR1        3 bed flat to rent   
2                  Downham Way, Bromley BR1        2 bed flat to rent   
3            Downham Way, Bromley, Kent BR1        2 bed flat to rent   
4  Downham Way, Bromley, Greater London BR1        2 bed flat to rent   

   Bedrooms  Bathrooms       Price   
0         1          1  £1,450 pcm  \
1         3          1  £2,250 pcm   
2         2          1  £1,700 pcm   
3         2          1  £1,300 pcm   
4         2          1  £1,250 pcm   

                                         Description   
0  Lawson S Stone are proud to present this amazi...  \
1  **two balconies & parking** inclusive of water...   
2  Two bedroom second floor apartment located on ...   
3  Recently refurbished 2 bedroom flat available ...   
4  Newly refurbished 2 bedroom flat available to ...   

                   Listing_Date                      Agent Listing_Source   
0       Listed on 25th May 2023             Lawson S Stone         Zoopla  \
1   Available from 1st Jun 2023          CKB Estate Agents         Zoopla   
2       Listed on 25th May 2023   Key Property Consultants         Zoopla   
3  Available from 10th Jul 2023  Daniels Property Services         Zoopla   
4       Listed on 25th May 2023  Daniels Property Services         Zoopla   

                                         listing_URL  
0  https://www.zoopla.co.uk/to-rent/property/br1/...  
1  https://www.zoopla.co.uk/to-rent/property/br1/...  
2  https://www.zoopla.co.uk/to-rent/property/br1/...  
3  https://www.zoopla.co.uk/to-rent/property/br1/...  
4  https://www.zoopla.co.uk/to-rent/property/br1/...

### i) Cleaning the Listing_Date column

We will attempt to clean the Listing_Date to remove the 'Listed on' string attached to the listed date

In [344]:
# Strip "Listed on" from Date values while preserving 
zp1['Listing_Date'] = zp1['Listing_Date'].str.replace('Listed on', '').str.strip()
zp1['Listing_Date'] = zp1['Listing_Date'].str.replace('Available from', '').str.strip()

In [345]:
zp1.head()

Unique_Id Location Tansaction_Type   
0  BR1R00001ZP      BR1            Rent  \
1  BR1R00002ZP      BR1            Rent   
2  BR1R00003ZP      BR1            Rent   
3  BR1R00004ZP      BR1            Rent   
4  BR1R00005ZP      BR1            Rent   

                              Property_Type                   Address   
0                    Park Road, Bromley BR1  1 bed maisonette to rent  \
1                Highland Road, Bromley BR1        3 bed flat to rent   
2                  Downham Way, Bromley BR1        2 bed flat to rent   
3            Downham Way, Bromley, Kent BR1        2 bed flat to rent   
4  Downham Way, Bromley, Greater London BR1        2 bed flat to rent   

   Bedrooms  Bathrooms       Price   
0         1          1  £1,450 pcm  \
1         3          1  £2,250 pcm   
2         2          1  £1,700 pcm   
3         2          1  £1,300 pcm   
4         2          1  £1,250 pcm   

                                         Description   Listing_Date   
0  Lawson S Stone are proud to present this amazi...  25th May 2023  \
1  **two balconies & parking** inclusive of water...   1st Jun 2023   
2  Two bedroom second floor apartment located on ...  25th May 2023   
3  Recently refurbished 2 bedroom flat available ...  10th Jul 2023   
4  Newly refurbished 2 bedroom flat available to ...  25th May 2023   

                       Agent Listing_Source   
0             Lawson S Stone         Zoopla  \
1          CKB Estate Agents         Zoopla   
2   Key Property Consultants         Zoopla   
3  Daniels Property Services         Zoopla   
4  Daniels Property Services         Zoopla   

                                         listing_URL  
0  https://www.zoopla.co.uk/to-rent/property/br1/...  
1  https://www.zoopla.co.uk/to-rent/property/br1/...  
2  https://www.zoopla.co.uk/to-rent/property/br1/...  
3  https://www.zoopla.co.uk/to-rent/property/br1/...  
4  https://www.zoopla.co.uk/to-rent/property/br1/...

Now, we will convert the Listing dates into the format dd/mm/yyyy, for uniformity with the other datasets

In [346]:
# Convert the date format to 'dd/mm/yyyy'
zp1['Listing_Date'] = zp1['Listing_Date'].str.replace('th', '').str.replace('st', '').str.replace('nd', '').str.replace('rd', '').str.replace(' ', '/')

# Convert the month name to the month number
try: 
    zp1['Listing_Date'] = pd.to_datetime(zp1['Listing_Date'], format='%d/%B/%Y').dt.strftime('%d/%m/%Y')
except ValueError:
    pass

# Convert the month name to the month number
try: 
    # Update the month name to month number conversion
    zp1['Listing_Date'] = zp1['Listing_Date'].str.replace('/Jan/', '/01/').str.replace('/Feb/', '/02/').str.replace('/Mar/', '/03/').str.replace('/Apr/', '/04/').str.replace('/May/', '/05/').str.replace('/Jun/', '/06/').str.replace('/Jul/', '/07/').str.replace('/Aug/', '/08/').str.replace('/Sep/', '/09/').str.replace('/Oct/', '/10/').str.replace('/Nov/', '/11/').str.replace('/Dec/', '/12/')
except ValueError:
    pass


In [347]:
zp1.head()

Unique_Id Location Tansaction_Type   
0  BR1R00001ZP      BR1            Rent  \
1  BR1R00002ZP      BR1            Rent   
2  BR1R00003ZP      BR1            Rent   
3  BR1R00004ZP      BR1            Rent   
4  BR1R00005ZP      BR1            Rent   

                              Property_Type                   Address   
0                    Park Road, Bromley BR1  1 bed maisonette to rent  \
1                Highland Road, Bromley BR1        3 bed flat to rent   
2                  Downham Way, Bromley BR1        2 bed flat to rent   
3            Downham Way, Bromley, Kent BR1        2 bed flat to rent   
4  Downham Way, Bromley, Greater London BR1        2 bed flat to rent   

   Bedrooms  Bathrooms       Price   
0         1          1  £1,450 pcm  \
1         3          1  £2,250 pcm   
2         2          1  £1,700 pcm   
3         2          1  £1,300 pcm   
4         2          1  £1,250 pcm   

                                         Description Listing_Date   
0  Lawson S Stone are proud to present this amazi...   25/05/2023  \
1  **two balconies & parking** inclusive of water...    1/06/2023   
2  Two bedroom second floor apartment located on ...   25/05/2023   
3  Recently refurbished 2 bedroom flat available ...   10/07/2023   
4  Newly refurbished 2 bedroom flat available to ...   25/05/2023   

                       Agent Listing_Source   
0             Lawson S Stone         Zoopla  \
1          CKB Estate Agents         Zoopla   
2   Key Property Consultants         Zoopla   
3  Daniels Property Services         Zoopla   
4  Daniels Property Services         Zoopla   

                                         listing_URL  
0  https://www.zoopla.co.uk/to-rent/property/br1/...  
1  https://www.zoopla.co.uk/to-rent/property/br1/...  
2  https://www.zoopla.co.uk/to-rent/property/br1/...  
3  https://www.zoopla.co.uk/to-rent/property/br1/...  
4  https://www.zoopla.co.uk/to-rent/property/br1/...

Convert any non-date values in the Listed_Date column to 'Not Listed'

In [348]:
# Convert the column to string type
zp1['Listing_Date'] = zp1['Listing_Date'].astype(str)

# Check if the value matches the date format
mask = zp1['Listing_Date'].str.match(r'\d{2}/\d{2}/\d{4}')

# Replace non-matching values with 'Not Listed'
zp1.loc[~mask, 'Listing_Date'] = 'Not Listed'

In [349]:
zp1.Listing_Date.value_counts()

Listing_Date
Not Listed    8722
23/05/2023     877
22/05/2023     757
18/05/2023     501
17/05/2023     487
              ... 
17/05/2024       1
28/06/2024       1
18/05/2024       1
21/01/2023       1
18/07/2022       1
Name: count, Length: 467, dtype: int64

Great.

<h1 style='background-color: Indigo; padding: 10px; color: white; text-align: center;'>DATA MERGING </h1>

Here, we will merge the three dataset for sales and rent properties separately before combining them into a master dataset

<h1 style='background-color: Indigo; padding: 10px; color: white'> Sales data </h1>

In [124]:
# Checking the shape of the three sales dataset
rm.shape, om.shape, zp.shape

((49641, 13), (14790, 13), (49494, 13))

In [138]:
# Total number of available columns
rm.shape[0] + om.shape[0]+ zp.shape[0]

113925

The total number of available properties is 113,925

We will now merge the three datasets and eliminate any duplicate entries. We will identify duplicates based on the following columns: 'Location', 'Transaction_Type', 'Property_Type', 'Address', 'Bedrooms', 'Bathrooms', 'Price', 'Description', and 'Agent'.

We will exclude the 'unique_id', 'listing_source', and 'source_url' columns from the duplicate check as these features are specific to each website.
    

In [146]:
# Concatenate the datasets vertically
merged = pd.concat([rm, om, zp], ignore_index=True)


# Specify the columns to check for duplicates
duplicate_cols = ['Location', 'Tansaction_Type', 'Property_Type', 'Address', 
                  'Bedrooms', 'Bathrooms', 'Price', 
                  'Description', 'Agent']

# Remove duplicates based on the specified columns
merged_sales = merged.drop_duplicates(subset=duplicate_cols)

# Reset the index of the merged DataFrame
merged_sales.reset_index(drop=True, inplace=True)




In [147]:
merged_sales.shape

(112978, 13)

After dropping the dublicates the number of available properties is 112,978 (947 Properties dropped)

In [143]:
# Saving to file
merged_sales.to_csv('sales_merged_test.csv')

<h1 style='background-color: Indigo; padding: 10px; color: white'> Rent data </h1>

In [352]:
# Checking the shape of the three sales dataset
rm1.shape, om1.shape, zp1.shape

((21126, 13), (22516, 13), (21873, 13))

In [353]:
# Total number of available columns
rm1.shape[0] + om1.shape[0]+ zp1.shape[0]

65515

We will go through the same process lke we did with the sales data

In [354]:
# Concatenate the datasets vertically
merged = pd.concat([rm1, om1, zp1], ignore_index=True)


# Specify the columns to check for duplicates
duplicate_cols = ['Location', 'Tansaction_Type', 'Property_Type', 'Address', 
                  'Bedrooms', 'Bathrooms', 'Price', 
                  'Description', 'Agent']

# Remove duplicates based on the specified columns
merged_rent = merged.drop_duplicates(subset=duplicate_cols)

# Reset the index of the merged DataFrame
merged_rent.reset_index(drop=True, inplace=True)




In [355]:
merged_rent.shape

(62429, 13)

In [356]:
65515 -62429

3086

After dropping the dublicates the number of available properties is 65,515 (3086 Properties dropped)

In [357]:
# Saving to file
merged_rent.to_csv('rent_merged_test.csv')